hazelcast/src/main/java/com/hazelcast/datalink/JdbcDataLink.java

```diff
@@ -0,0 +1,183 @@
+     * the pooledDataSource
+     *
+     * We increase the counter both in DataLinkService and DataLink to avoid race condition when
+     * - a JdbcDataLink is retrieved from the service
+     * - the DataLink is removed and the pooledDataSource is closed
+     * - a closed pooledDataSource is returned
+     */
+    protected final ReferenceCounter refCounter;
+    protected final DataLinkConfig config;
+
+    protected CloseableDataSource pooledDataSource;
+    protected DataSource singleUseDataSource;
+
+    public JdbcDataLink(DataLinkConfig config) {
+        this.refCounter = new ReferenceCounter(() -> {
+            destroy();
+            return null;
+        });
+        this.config = config;
+        this.pooledDataSource = createHikariDataSource();
+        this.singleUseDataSource = createSingleUseDataSource();
+    }
+
+    private DataSource createSingleUseDataSource() {
+        Properties properties = config.getProperties();
+        String jdbcUrl = properties.getProperty("jdbcUrl");
+        return new DataSourceFromConnectionSupplier(
+                () -> {
+                    try {
+                        Properties connectionProps = new Properties();
+                        properties.entrySet().stream().filter(e -> !"jdbcUrl".equals(e.getKey()))
+                                  .forEach(e -> connectionProps.put(e.getKey(), e.getValue()));
+                        return DriverManager.getConnection(jdbcUrl, connectionProps);
+                    } catch (SQLException e) {
+                        throw new HazelcastException("Could not create a new connection", e);
+                    }
+                }
+        );
+    }
+
+    @Override
+    public String getName() {
+        return config.getName();
+    }
+
+    @Override
+    public List<Resource> listResources() {
+        try (Connection connection = singleUseDataSource.getConnection()) {
+            DatabaseMetaData metaData = connection.getMetaData();
+            String[] types = {"TABLE", "VIEW"};
+            //Retrieving the columns in the database
+            ResultSet tables = metaData.getTables(null, null, "%", types);
+            List<Resource> result = new ArrayList<>();
+            while (tables.next()) {
+                result.add(new Resource(
+                        tables.getString("TABLE_TYPE"),
+                        tables.getString("TABLE_SCHEM") + "." + tables.getString("TABLE_NAME")
+                ));
+            }
+            return result;
+        } catch (Exception e) {
+            throw new HazelcastException("Could not read resources for DataLink " + config.getName());
+        }
+    }
+
+    @Override
+    public DataLinkConfig getConfig() {
+        return config;
+    }
+
+    public DataSource pooledDataSource() {
+        retain();
+        return pooledDataSource;
+    }
+
+    public DataSource singleUseDataSource() {
+        return singleUseDataSource;
+    }
+
+    public DataSource getDataSource() {
```

Comment:
I think most of the implementations will have similar method. Maybe push it to the interface with more general name, like `getLink()`?

Reply:
This method slightly deviates from the TDD, which defines the two methods above: `pooledDataSource` and `singleUseDataSource`.

In general, this will be too implementation specific depending on the 3rd party client/driver - e.g. for Kafka there will be a shared producer, but only a single use consumer etc.. 

The DataLinkService gives you API to get the implementation - e.g. concrete JdbcDataLink instead of DataLink so you can still conveniently get, what you need.